daceml/distributed/schedule.py

```diff
@@ -0,0 +1,339 @@
+
+    * Detect and handle local write-conflicts
+    * Handle views correctly
+    * Insert rank-local arrays and reroute the map edges to go to the local
+      views
+    """
+    me, mx = map_nodes
+
+    # Compute tiled map range
+    new_range = compute_tiled_map_range(me.map, num_blocks)
+
+    rank_variables, (input_subsets,
+                     output_subsets) = propagate_rank_local_subsets(
+                         sdfg, state, map_nodes, num_blocks)
+
+    # set the new range
+    me.map.range = new_range
+
+    to_iter = itertools.chain(
+        zip(input_subsets.items(), itertools.repeat(True)),
+        zip(output_subsets.items(), itertools.repeat(False)))
+
+    result_read: GlobalToLocal = []
+    result_write: GlobalToLocal = []
+    for (arr_name, new_subset), is_input in to_iter:
+
+        # Determine the outer edge
+        outer_edges = state.in_edges(me) if is_input else state.out_edges(mx)
+        outer_edge_cands = [
+            edge for edge in outer_edges if edge.data.data == arr_name
+        ]
+        if len(outer_edge_cands) > 1:
+            # FIXME this could be supported using a preprocessing
+            # transformation
+            raise NotImplementedError(
+                "Multiple outer edges to one array not implemented")
+        elif len(outer_edge_cands) == 0:
+            raise ValueError(f"No outer edge to {arr_name}")
+        outer_edge = outer_edge_cands[0]
+
+        global_name = outer_edge.data.data
+        global_node = outer_edge.src if is_input else outer_edge.dst
+
+        if not isinstance(global_node, nodes.AccessNode):
+            # FIXME tasklets should be replicated for each rank
+            raise NotImplementedError("Cannot handle non-access nodes yet")
+        elif isinstance(global_node.desc(sdfg), data.View):
+            raise NotImplementedError("Cannot handle views yet")
+
+        # Create the rank-local array
+        local_name, _ = sdfg.add_transient(
+            name="local_" + global_name,
+            shape=[
+                symbolic.overapproximate(r).simplify()
+                for r in new_subset.size_exact()
+            ],
+            dtype=sdfg.arrays[global_name].dtype,
+            find_new_name=True)
+
+        local_node = state.add_access(local_name)
+        if is_input:
+            result_read.append((global_node, local_node, new_subset))
+        else:
+            result_write.append((global_node, local_node, new_subset))
+
+        if is_input:
+            redirect_args = dict(new_src=local_node)
+        else:
+            redirect_args = dict(new_dst=local_node)
+
+        new_edge = xfh.redirect_edge(state,
+                                     outer_edge,
+                                     new_data=local_name,
+                                     **redirect_args)
+        new_edge.data.subset = new_subset
+
+    return rank_variables, result_read, result_write
+
+
+def lower(sdfg: SDFG, schedule: DistributedSchedule):
```

Comment:
A bit of an undocumented and uneventful name for what is the driver function of this file :)